File: rust/src/matrix/matrix.rs
pub struct Matrix {
  data: Vec<Vec<f32>>,
  rows: usize,
  cols: usize
}

impl Matrix {
  pub fn new(rows: usize, cols: usize) -> Matrix {
    Matrix {
      data: vec![vec![0.0; cols]; rows],
      rows,
      cols
    }
  }

  pub fn get(&self, row: usize, col: usize) -> f32 {
    self.data[row][col]
  }

  pub fn set(&mut self, row: usize, col: usize, value: f32) {
    self.data[row][col] = value;
  }

  pub fn rows(&self) -> usize {
    self.rows
  }

  pub fn cols(&self) -> usize {
    self.cols
  }

  pub fn to_array(&self) -> Vec<f32> {
    if self.rows() == 1 {
      return self.data[0].to_vec();
    } else if self.cols() == 1 {
      let mut result = vec![];
      for row in 0..self.rows() {
        result.push(self.get(row, 0));
      }
      return result;
    } else {
      panic!("Cannot convert matrix to array");
    }
  }

  pub fn from_array_row(array: &Vec<f32>) -> Matrix {
    let mut matrix = Matrix::new(1, array.len());
    for i in 0..array.len() {
      matrix.set(0, i, array[i]);
    }
    matrix
  }

  pub fn from_array_col(array: &Vec<f32>) -> Matrix {
    let mut matrix = Matrix::new(array.len(), 1);
    for i in 0..array.len() {
      matrix.data[i][0] = array[i];
    }
    matrix
  }
}

impl std::fmt::Display for Matrix {
  fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
    for row in 0..self.rows() {
      for col in 0..self.cols() {
        write!(f, "{:.5} ", self.get(row, col))?;
      }
      write!(f, "\n")?;
    }
    Ok(())
  }
}

File: server/src/nn_lib/Matrix.ts
import { gridToArray } from "./matrix_math.js"

export class Matrix {
  data: Float32Array
  rows: number
  cols: number

  constructor(rows: number, cols: number, data?: Float32Array) {
    this.rows = rows
    this.cols = cols
    this.data = data || new Float32Array(rows * cols * 4)
  }

  set(row: number, col: number, value: number) {
    this.data[gridToArray(row, col, this.rows, this.cols, 4)] = value
  }

  get(row: number, col: number): number {
    return this.data[gridToArray(row, col, this.rows, this.cols, 4)]
  }

  toArray(): number[] {
    if(this.rows == 1) {
      let result: number[] = []
      for(let i = 0; i < this.cols; i++) {
        result[i] = this.get(0, i)
      }
      return result

    } else if(this.cols == 1) {
      let result = []
      for(let row = 0; row < this.rows; row ++) {
        result.push(this.get(row, 0))
      }
      return result;
    } else {
      throw "Cannot convert matrix to array"
    }
  }

  static fromArrayRow(array: number[]): Matrix {
    let matrix = new Matrix(1, array.length)
    for(let i = 0; i < array.length; i++) {
      matrix.set(0, i, array[i]);
    }
    return matrix
  }

  static fromArrayCol(array: number[]): Matrix {
    let matrix = new Matrix(array.length, 1)
    for(let i = 0; i < array.length; i++) {
      matrix.set(i, 0, array[i]);
    }
    return matrix
  } 

  toString() {
    let result = ""
    for(let row = 0; row < this.rows; row ++) {
      for(let col = 0; col < this.cols; col ++) {
        result += this.get(row, col).toFixed(2) + " "
      }
      result += "\n"
    }
    return result
  }
}